### ISSUE TITLE
Wrong primary monitor when turning surround off with DisplayMagician.

### ISSUE THREAD

**Issue by Chrius** [terrymacdonald/DisplayMagician#117]
Every time I swap from my surround profile to the normal (3 monitor) one, windows sets my right monitor as primary instead of the center one.
Then I have to go to the Nvidia Control panel and uncheck the "Span Displays with Surround" checkbox and everything goes back to how it should be. 

The monitors actually show up as 3 separate monitors with settings for each one (instead of one giant monitor running at 5760x1080), but with the wrong one as primary.
After unchecking the Surround checkbox, it seems to just remove a flag somewhere that it's using it, because the monitors don't actually reset (like they do when you normally enable or disable surround), just the primary monitor gets set as the correct one again.

**Comment by terrymacdonald** [comment]
Hmmmm. It sounds like an NVIDIA driver problem, but can you please generate
a Support ZIP File from the settings window? Then upload it here through
your web browser.

I'll take a look.

Thanks
Terry

**Comment by terrymacdonald** [comment]
@Chrius Also please provide more information about what NVIDIA driver you are on, what monitors you are using, and what types of connections do you use to each of them (e.g. displayport, HDMI, DVI-D, etc)

Thanks
Terry

**Comment by terrymacdonald** [comment]
@Chrius also what OS you are running, the model of your video card, if it is a standalone video card or in a laptop. Oh, and if there is anything else connected to the PC (like HTC Vive or something), or if you are running other software that changes the windows taskbar, or that adds different displays (like Superdisplay or SpaceDesk).

Thanks
Terry

**Comment by terrymacdonald** [comment]
@Chrius One REALLY last thing. What version of DisplayMagician are you running? v2.3.1.60?

**Comment by Chrius** [comment]
[DisplayMagician-Support-20220606-0818.zip](https://github.com/terrymacdonald/DisplayMagician/files/8843996/DisplayMagician-Support-20220606-0818.zip)

I'm using the 512.95 driver on a GTX 1660 (Standalone) on Windows 11 (21H2) with DisplayMagician version 2.3.1.60. 
The monitors are 2 LG Ultragear 24" and 1 LG Ultragear 27" (all at 1080p).
The 27" is connected through Displayport and the other 2 with HDMI and a DVI-D to HDMI cable (my card has only 1DP, 1 HDMI and 1 DVI-D connection).
On my default config I'm using the 27" at 240Hz and the 24"s at 120Hz. Surround I have at 60Hz atm.

I don't have anything else connected to any video connections, and I'm not using any other software that changes the displays.

Before I started using DisplayMagician I was using StartAllBack, but I had some problems where I could't manually disable Surround. So I uninstalled StartAllBack and installed Start11. I also had to run DDU to completely remove the Nvidia driver and reinstall to fix the earlier surround problem.

And btw, I also noticed that when I switch out of Surround, all my monitors stay at 60Hz, even though they are all separated, and unchecking the Surround checkbox also makes them return to the correct refresh rate.

**Comment by terrymacdonald** [comment]
@Chrius ok - I think this is actually caused by one of the errors that is fixed in the upcoming DisplayMagician v2.4.0. I will give you a download link, but please note this is a 'test' version, and you will need to delete and recreate your Display Profiles again when the final DisplayMagician version is released.

[DisplayMagicianSetup-v2.4.0.10.zip](https://github.com/terrymacdonald/DisplayMagician/files/8835910/DisplayMagicianSetup-v2.4.0.10.zip)

Thanks
Terry

**Comment by terrymacdonald** [comment]
I'm going to close this issue for now as I believe it will be fixed by the DisplayMagician test version I just linked above. If you still have issues after you have installed v2.4.0.10 then please reopen this issue and we'll keep on testing.

Thanks
Terry

**Comment by terrymacdonald** [comment]
Actually @Chrius I've just created a new issue #119 for this exact issue, as someone else tested with DisplayMagician v2.4.0.10 and had the same problem that you described.

Can you please confirm that you also had the primary display not being set issue on DisplayMagician v2.3.1.60? 

Thanks
Terry

**Comment by terrymacdonald** [comment]
@Chrius , I've found a couple of things that seem to be the cause of the error you got. 

1. It looks like NVIDIA have made their NvAPI_Mosaic_SetDisplayGrids function only work with a SET_TOPOLOGY_FLAG value of NONE in the latest version of the driver. All other flag values results in a Crash to Desktop (CTD) :(. I have removed the other flags I used so that the driver will keep working.

2. Patch adapter ids (which patches Windows adapter ids so they work after a reboot) had a bug caused in a recent update. This would only update the first adapter seen, and not any others. This meant that a config wouldn't work after a reboot.

I have also updated the UI window placement when applying a profile. Because the screen coordinates change when applying a new display layout, the UI used to sometimes go off screen, and it was a pain to try and move it to the right place so you could see it again. I have changed that so that when a display profile is applied it will move the UI to the middle of the primary monitor, and will also then center the Display Profile screen. This should make DisplayMagician far easier to use.

To install DisplayMagician v2.4.0.26, please download this zip file and install the exe inside it. Please test the software and make sure that it now swaps display profiles without any errors: 
[DisplayMagicianSetup-v2.4.0.26.zip](https://github.com/terrymacdonald/DisplayMagician/files/8869771/DisplayMagicianSetup-v2.4.0.26.zip)

Please note, as DM v2.4.0.26 has some new display profile features it stores (Windows DPI Scaling and NVIDIA 3D Settings), old Display Profiles won't work. DM v2.4.0.26 will ask you to recreate your Display Profiles. Please do that, then update your Game Shortcuts to use those new Display Profiles.

Please test DM v2.4.0.26 and make sure you can swamp display profiles without any errors. I'm crossing my fingers that I've finally sorted the errors you've experienced 🤞.

Thanks
Terry

**Comment by Chrius** [comment]
> Can you please confirm that you also had the primary display not being set issue on DisplayMagician v2.3.1.60?
Yeah

It's weird, with 2.4.0.10 (the previous version you provided) it seemed to be working correctly (at least with the limited testing I did). 
I noticed it was still not unchecking the Surround box on the Nvidia control panel, but the monitors were working as they were supposed to and each with the correct refresh rate.

But when I come home from work I'll test the new version.

**Comment by Chrius** [comment]
Ok, with 2.4.0.26 it's working the same way as it was with 2.4.0.10 for me.

The settings are working as they should, but the surround checkbox is not being unchecked (this isn't an issue anymore since the first 2.4 release, because even though it's checked, all settings are working as they should).

**Comment by terrymacdonald** [comment]
Hi @Chrius,

The control panel can't be unchecked by DisplayMagician. That is a
limitation of the NVIDIA control panel that can't be adjusted. The NVIDIA
control panel keeps track of its own UI state, and it expects that the only
way to change surround is via the control panel... This means that the
NVIDIA programmers didn't make their software check to make sure that
Surround was on our not. It just keeps the last setting you set.

The important think it's that it has no bearing on the display layout at
all. In all my testing, I could make any changes, and the surround checkbox
wouldn't change, yet when I used the NvAPI DLL to talk directly to the
video card the settings were changed by DisplayMagician correctly.

So, in other words, the checkbox starting checked isn't a bug, it's a
feature 😜. And nothing I can do I'm afraid.

I will close this issue then if it's working. Thanks for logging the issue,
and of you find anything else broken please log another issue.

Thanks
Terry
